contracts/DKVDaggerHashimoto.sol

```diff
@@ -212,7 +216,66 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
             uint256 parent = uint256(hash0) % rows;
             uint256 kvIdx = parent + (startShardId << shardEntryBits);
             bytes memory data = maskedData[i];
-            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data), "invalid access proof");
+            require(systemContract.checkDaggerData(kvIdx, kvMap[idxMap[kvIdx]].hash, data),
+                    "invalid access proof");
+
+            assembly {
+                mstore(data, hash0)
+                hash0 := keccak256(data, add(maxKvSize, 0x20))
+                mstore(data, maxKvSize)
+            }
+        }
+        return hash0;
+    }
+
+    /* Helper function to avoid stack to deep error */
+    function _generateKVIdx(
+        bytes32 hash0,
+        uint256 startShardId,
+        uint256 shardLenBits
+    ) internal view returns (uint256) {
+        uint256 rows = 1 << (shardEntryBits + shardLenBits);
+        uint256 parent = uint256(hash0) % rows;
+        uint256 kvIdx = parent + (startShardId << shardEntryBits);
+        return kvIdx;
+    }
+
+    function _generateChunkIdx(
+        bytes32 hash0
+    ) internal view returns (uint256) {
+        require((CHUNK_SIZE != 0) && ((CHUNK_SIZE & (CHUNK_SIZE - 1)) == 0), "CHUNK_SIZE invalid");
```

Comment:
We should not check 2**n of CHUNK_SIZE here - it should be in the ctor code so that we don't need to pay the gas per call.

Reply:
Make senses.